pkg/cloudprovider/aws.go

```diff
@@ -1,37 +1,247 @@
 package cloudprovider
 
 import (
+	"fmt"
 	"net"
 
+	awsapi "github.com/aws/aws-sdk-go/aws"
+	awscredentials "github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
 	corev1 "k8s.io/api/core/v1"
+	utilnet "k8s.io/utils/net"
 )
 
 const (
 	aws = "aws"
 )
 
-// AWS implements the API wrapper for talking
-// to the AWS cloud API
+var (
+	awsIPv4FilterKey = "network-interface.addresses.private-ip-address"
+	awsIPv6FilterKey = "network-interface.ipv6-addresses.ipv6-address"
+)
+
+// AWS implements the API wrapper for talking to the AWS cloud API
 type AWS struct {
-	cloud CloudProvider
+	CloudProvider
+	region string
+	client *ec2.EC2
+}
+
+// AWSWaitInput is the required input for the AWS EC2 Wait API call (WaitUntilInstanceRunning).
+// Unfortunately the API only handles equality assertion: so on delete we can't
+// specify and assert that the IP which is being removed is completely removed,
+// we are forced to do the inverse, i.e: assert that all IPs except the IP being
+// removed are there (which should anyways be the case). Hence we might risk
+// having a very small time window where the state is incorrect. This can be
+// problematic for update operation. The assumption here is that AWS EC2 will be
+// able to handle this. Unfortunately, my research showed that there is no other
+// API call that we can use to avoid this.
+type AWSWaitInput struct {
+	instanceID *string
+	ips        []*string
 }
 
 func (a *AWS) initCredentials() error {
+	accessKey, err := a.readSecretData("aws_access_key_id")
+	if err != nil {
+		return err
+	}
+	secretKey, err := a.readSecretData("aws_secret_access_key")
+	if err != nil {
+		return err
+	}
+	mySession := session.Must(session.NewSession())
+	a.client = ec2.New(mySession, awsapi.NewConfig().WithCredentials(awscredentials.NewStaticCredentials(accessKey, secretKey, "")).WithRegion(a.region))
 	return nil
 }
 
 func (a *AWS) AssignPrivateIP(ip net.IP, node *corev1.Node) (interface{}, error) {
-	return nil, nil
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return nil, err
+	}
+	addIP := ip.String()
+	keepIPs := []*string{}
+	if utilnet.IsIPv6(ip) {
+		for _, assignedIPv6 := range instance.NetworkInterfaces[0].Ipv6Addresses {
```

Comment:
Is it possible that NetworkInterfaces is an empty list? If so, this will panic

Reply:
I don't think so. The testing and research I've made on AWS seems to indicate that a VM will always have at least one network interface. I am adding a check anyways however.